docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
+    
+### Security Context Provider
+
+The Kubelet will have an interface that points to a `SecurityContextProvider`. The `SecurityContextProvider` is invoked before creating and running a given container:
+
+```go
+type SecurityContextProvider interface {
+    // ModifyContainerConfig is called before the Docker createContainer call.
+    // The security context provider can make changes to the Config with which
+    // the container is created.
+    // An error is returned if it's not possible to secure the container as 
+    // requested with a security context. 
+	ModifyContainerConfig(pod *api.BoundPod, container *api.Container, config *docker.Config) error
+	
+	// ModifyHostConfig is called before the Docker runContainer call.
+	// The security context provider can make changes to the HostConfig, affecting
+	// security options, whether the container is privileged, volume binds, etc.
+	// An error is returned if it's not possible to secure the container as requested 
+    // with a security context. 
+	ModifyHostConfig(pod *api.BoundPod, container *api.Container, hostConfig *docker.HostConfig)
+}
+```
+If the value of the SecurityContextProvider field on the Kubelet is nil, the kubelet will create and run the container as it does today.   
+
+### Security Context
+
+A security context has a 1:1 correspondence to a service account and it can be included as
+part of the service account resource. Following is an example of an initial implementation:
+
+```go
+type SecurityContext struct {
+    // user is the uid to use when running the container
+	User int
+	
+	// allowPrivileged indicates whether this context allows privileged mode containers
+	AllowPrivileged bool
+	
+	// allowedVolumeTypes lists the types of volumes that a container can bind
+	AllowedVolumeTypes []string
+	
+	// addCapabilities is the list of Linux kernel capabilities to add
+	AddCapabilities []string
+	
+	// removeCapabilities is the list of Linux kernel capabilities to remove
+	RemoveCapabilities []string
+	
+	// SELinux specific settings (optional)
+	SELinux *SELinuxContext
+	
+	// AppArmor specific settings (optional)
+	AppArmor *AppArmorContext
+	
+	// FUTURE:
+	// With Linux user namespace support, it should be possible to map
+	// a range of container uids/gids to arbitrary host uids/gids
+	// UserMappings []IDMapping
+	// GroupMappings []IDMapping
+}
+
+type SELinuxContext struct {
+    // MCS label/SELinux level to run the container under
+    Level string
+    
+    // SELinux type label for container processes
+    Type  string    
+    
+    // FUTURE:
+    // LabelVolumeMountsExclusive []Volume
+    // LabelVolumeMountsShared    []Volume
+}
+
+type AppArmorContext struct {
+	// AppArmor profile
+	Profile string
+}
```

Comment:
----- Original Message -----
> > +type SELinuxContext struct {
> > +    // MCS label/SELinux level to run the container under
> > +    Level string
> > +
> > +    // SELinux type label for container processes
> > +    Type  string
> > +
> > +    // FUTURE:
> > +    // LabelVolumeMountsExclusive []Volume
> > +    // LabelVolumeMountsShared    []Volume
> > +}
> > +
> > +type AppArmorContext struct {
> > +	// AppArmor profile
> > +	Profile string
> > +}
> 
> I agree that we would want to use all the mechanisms mentioned above
> (capabilities, MCS labels, apparmor profiles) if available.  And that the
> initial implementation should use these, since RedHat has so much expertise
> with these.
> 
> At the same time, it is very much tied to a specific implementation.  This
> makes it harder for users to understand so much detail; harder to drop in
> alternative implementations should we ever want to do that.  Examples of
> different implementations:  some company might use grsecurity and Pax
> already. (I don't but someone might); some hosting provider might write a
> different implementation that has similar effect (I can see us doing that).
> 
> So, can you think of a way to divide this up into two layers: one that is a
> core API object that expresses intent, and another, which implements the
> intent?

At a minimum, anything that is not 100% all Linuxes should be an extension plugin (or a default extension).  No disagreement from me.

> 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3910/files#r23858436